lib/Feature/FeatureModel.cpp

```diff
@@ -0,0 +1,148 @@
+#include "vara/Feature/FeatureModel.h"
+
+namespace vara::feature {
+void FeatureModel::dump() const {
+  for (const auto &F : Features) {
+    llvm::outs() << "{\n";
+    F.second->dump();
+    llvm::outs() << "}\n";
+  }
+  llvm::outs() << '\n';
+}
+
+bool FeatureModel::addFeature(std::unique_ptr<Feature> F) {
+  std::string K = F->getName();
+  if (!Features.try_emplace(K, std::move(F)).second) {
+    return false;
+  }
+  Feature *V = Features[K].get();
+  for (auto *C : V->children()) {
+    C->setParent(V);
+  }
+  if (V->isRoot()) {
+    if (*Root->getParent() == *V) {
+      Root = V;
+    } else {
+      V->setParent(Root);
+    }
+  } else {
+    V->getParent()->addChild(V);
+  }
+  OrderedFeatures.insert(V);
+  return true;
+}
+
+void FeatureModelBuilder::buildConstraints() {
+  for (const auto &F : Features.keys()) {
+    for (const auto &E : Excludes[F]) {
+      Features[F]->addExclude(Features[E].get());
+    }
+  }
+
+  for (const auto &C : Constraints) {
```

Comment:
If we iterate over lists here we should use a better name than `C`. What is the actual type of `C`? maybe it could help to spell it out instead of auto.